### ISSUE TITLE
PREDICTIVE MAINTENANCE IN SOLAR POWER SYSTEMS

### ISSUE THREAD

**Issue by MG21102004** [MG21102004/GAJENDHIRAN_0E3D2C1C26C52FAA67241FE78344854B#1]
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

# Step 1: Simulate Solar Panel Sensor Data
def generate_solar_data(samples=1000):
    np.random.seed(42)
    
    # Normal conditions
    temp = np.random.normal(45, 5, samples)  # Temperature (°C)
    irradiance = np.random.normal(800, 50, samples)  # Solar irradiance (W/m²)
    voltage = np.random.normal(240, 2, samples)  # Voltage (V)
    current = np.random.normal(8, 0.5, samples)  # Current (A)
    
    # Introduce faults (~10% of data)
    fault_indices = np.random.choice(samples, size=int(samples*0.1), replace=False)
    temp[fault_indices] += np.random.normal(25, 5, len(fault_indices))  # Overheating
    voltage[fault_indices] *= 0.7  # Voltage drop
    current[fault_indices] *= 0.6  # Current drop
    
    # Labels (0 = Normal, 1 = Fault)
    labels = np.zeros(samples)
    labels[fault_indices] = 1
    
    return pd.DataFrame({
        'temperature': temp,
        'irradiance': irradiance,
        'voltage': voltage,
        'current': current,
        'fault': labels
    })

# Step 2: Feature Engineering
def add_features(df):
    df['power_output'] = df['voltage'] * df['current']  # P = V × I
    df['efficiency'] = df['power_output'] / df['irradiance']  # Efficiency metric
    df['temp_irradiance_ratio'] = df['temperature'] / df['irradiance']  # Thermal stress indicator
    return df

# Step 3: Train-Test Split & Model Training
def train_predictive_model(df):
    X = df.drop('fault', axis=1)
    y = df['fault']
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    
    # Pipeline: Scaling + Random Forest
    model = Pipeline([
        ('scaler', StandardScaler()),
        ('classifier', RandomForestClassifier(n_estimators=100, random_state=42))
    ])
    
    model.fit(X_train, y_train)
    return model, X_test, y_test

# Step 4: Evaluate Model
def evaluate_model(model, X_test, y_test):
    y_pred = model.predict(X_test)
    print("Accuracy:", accuracy_score(y_test, y_pred))
    print("\nClassification Report:\n", classification_report(y_test, y_pred))

# Main Execution
if __name__ == "__main__":
    print("Generating solar panel data...")
    solar_data = generate_solar_data(5000)
    
    print("Adding features...")
    solar_data = add_features(solar_data)
    
    print("Training model...")
    model, X_test, y_test = train_predictive_model(solar_data)
    
    print("\nModel Evaluation:")
    evaluate_model(model, X_test, y_test)
